### PR TITLE
Base setup for translating directly in the admin panel

### PR DESCRIPTION
**Pull request by Pierstoval**
This feature is clearly based on [@adrianolek/AOTranslationBundle](https://github.com/adrianolek/AOTranslationBundle)'s powerful translation panel right inside the toolbar and the Symfony profiler.

Ping @Rvanlaak : can you test this in one of your environment and tell me if this suits to you? :smile:

### CODE REVIEW

**Comment by Rvanlaak** [comment]
Great! will test it within the next couple of days

**Comment by Rvanlaak** [comment]
My first impression: great work! I've came across a couple of configuration bugs.

1) When users have defined the Doctrine ORM in their own way, the bundle should be added to the manager's mapping:

```yml
doctrine:
    orm:
        entity_managers:
            default:
                mappings:
                    OrbitaleTranslationBundle: ~
```

2) My advise would be to only add the bundle when env != prod
```php
if ($this->getEnvironment() !== 'prod') {
        $bundles[] = new Orbitale\Bundle\TranslationBundle\OrbitaleTranslationBundle();
}
```
3) The Profiler does capture all translations properly when I've configured my application as follows:

```yml
default_locale: "undefined"
    translator:
        fallback: "undefined"
```
But when I set the default and fallback, the Translator does not capture any keys. We're using .yml files for the translations:
```yml
default_locale: nl
    translator:
        fallback: en
```
The translations are properly shown, but the bundle does not seem to capture the .yml translations.

4) An extra feature that in my opinion is a must-have is to show all locales in the profiler translation page. This way I don't know where I should translate

**Comment by Rvanlaak** [comment]
ping @Pierstoval

**Comment by Pierstoval** [comment]
Hi @Rvanlaak ! Thanks for pinging me.
 
For the two first points, I think it's just a documentation issue, I'll add some infos in the readme file.
 
For point 3, it's linked to point 4 : I truely want to override Symfony's translator, so I think I'm gonna have to use compiler passes to override the default translator with Orbitale's one. After that, I even think that the listener can be removed.
 
What do you think ?

**Comment by Rvanlaak** [comment]
It seems to me the listener should be removed, right? It is something the developer has to manage in my opinion. If the locale isn't set properly, it should be fixed without this bundle having to fix it with the listener.

My scenario is:

1. Locale is configured as "undefined"
1. Request locale is set based on the locale that is saved in the User entity
2. Translator is initiated, based on Request locale, so is set to the User's locale

What I'd like to archieve is:
* To read all .yml translations from my current translation files into the translation database
* To configure the translation database across multiple local installations of my project
* To be able to see the DateTime when the translation was added
* To be able to see the DateTime when the translation was modified for the last time (and maybe a serialized version of the `$request->getUser()` that modified it)
* To translate based on the request (this issue)
* To add xlf notes while translating
* To finally dump all xlf files in `/app/Resources/translations` and check it in to git

You've already [configured a translation loader](https://github.com/Orbitale/TranslationBundle/blob/master/Resources/config/services.yml#L2-L7) for the `.db` filetype, but that won't work right? How do you solved this piece of documentation?

> If you're loading translations from a database, you'll still need a resource file, but it might either be blank or contain a little bit of information about loading those resources from the database. The file is key to trigger the load method on your custom loader.

So about the compiler passes, maybe [tagging your Translator](http://symfony.com/doc/current/reference/dic_tags.html#translation-loader) is the way to go?

**Comment by Pierstoval** [comment]
> * To read all .yml translations from my current translation files into the translation database

This might be possible with a new command inspired by the `debug:translation` one, but IMO this is not relevant for this bundle's main purpose.
As a matter of fact, this bundle is essentially made for apps where the translator is not the same as the developer, or when the translation is made outside the dev process.
 
The basic scenario is:

* I want to translate a new / an existing app that is untranslated, or partially
* Install this bundle
* All missing translations are registered in the database "on the fly", so I can keep developing and run unit tests on my app without problem
* Manage them in SonataAdmin or EasyAdmin or native's back-end, or with the profiler "on the fly" when I see missing translations in the web debug toolbar
* Export them in `app/Resources/translations`, mostly
* Then in production they're imported automatically from this directory and put in cache

The execution context with collecting datas in the profiler is a very great feature because you know exactly which page has to be translated, and you know the context (posted form, flash message, etc.).
But basically, the context *should* be obvious with 2 parameters: the source id , and the translation domain.
Having a `Accept` source in the `message` translation domain sounds like a bad practice to me, whether you are using Orbitale's bundle or not does not matter. Even if you perfectly know the context when using the profiler to translate, this kind of non-specific behavior might lead to issues one day.
Instead, having something like `button.general_sales_condition.accept` in the domain `pages.about` sounds relevant, coherent and correct, and you don't need any context information with this because they're explicitly specified.
 
If you want to create such command, feel free to do it, it might be an "extra" to this bundle, but should not represent a good practice anyway, except you *really* know what to do.
 
Also, what I don't like in this idea is that, with this method, even if there's a good possibility to change translations of dependencies (like FOSUserBundle for instance), the problem is, to extract translations from these bundles, you have to retrieve **all** translations. Imagine FOSUserBundle with its heavy translation file. Now imagine other big bundles that are fully translated, in many languages. It's too heavy, especially if you want to override only one message. 
Plus, the dev is meant to execute this command, and I want this bundle usable with the least possible dev intervention.
 
What I really need with this bundle is simply adding the db abstraction layer to Symfony, in the sweetest and the smoothest possible way so Symfony can continue to work without problem. That's precisely why the translator extends Symfony's one. You pointed many weaknesses of the bundle, and I'm grateful for that, we may work on solving this all :) 

> * To configure the translation database across multiple local installations of my project

You think having a single translation database for all your local apps? If you're using two connections, then it sounds legit to me to add something like what other bundles do, like a simple `db_connection` parameter or something.

> * To be able to see the DateTime when the translation was added
> * To be able to see the DateTime when the translation was modified for the last time (and maybe a serialized version of the `$request->getUser()` that modified it)

It would need `Timestampable` and `Blameable` behaviors, or at least rewrite them, and I don't either want to rewrite or add a dependency to this bundle, it would make it more complex.

> * To add xlf notes while translating

This will be the object of another issue :wink: 

> You've already [configured a translation loader](https://github.com/Orbitale/TranslationBundle/blob/master/Resources/config/services.yml#L2-L7) for the `.db` filetype, but that won't work right? How do you solved this piece of documentation?

It works "partially". In fact, the db loader has to work out-of-the-box of the basic Symfony's loader, and has to work with a static catalogue. Unless, it would cause errors when persisting missing translations, because a simple `render(controller())` can break the whole thing.
Imagine a website with pages and categories. I'd bet one'd put the categories inside a navigation bar, and one nice solution is to use a controller to render the menu right inside the `base.html.twig` template for instance. Then, the script process continues and you also have a summary of the categories inside the footer. You also have the category inside the page.
All these elements might be translated. And if you don't use the `Translatable` behavior, you might (even if it is not the best practice, it's a valid one) translate directly the output of these categories' names. It is really legit if these categories are *frozen* with fixtures and if they are not supposed to change.
So, we might see these categories more than thrice. And if they're translated, then, the translator will call the database thrice. One in the view rendered by the nav controller, another by the footer controller, and another by the main-requested controller. It means that 3 kernels will be created. So, 3 kernels mean 3 containers. And 3 containers mean 3 entity managers. Each one of them will call the `flushTranslations()` event when reaching the `kernel.terminate` event. 
And if the catalogue is not static, then it will have 3 times the same catalogue in the translator, so it means 3 database queries, and 3 flushes. But 2 of them will fail because it will try to persist items that already exist because of the first flush.
 
That's why I'm using a static catalogue. 
If you have a better solution, or if I'm wrong, please remind me, I made this bundle one year ago (IIRC), so things might have change since :)

> So about the compiler passes, maybe [tagging your Translator](http://symfony.com/doc/current/reference/dic_tags.html#translation-loader) is the way to go? 
 
Maybe, yes, I think that overriding the native translator is the best solution, because it would be triggered when using `$this->container->get('translator')` in a controller for instance.
 
Thanks a lot for supporting this bundle!

**Comment by Rvanlaak** [comment]
> This might be possible with a new command inspired by the debug:translation one, but IMO this is not relevant for this bundle's main purpose.
> As a matter of fact, this bundle is essentially made for apps where the translator is not the same as the developer, or when the translation is made outside the dev process

You're right about this, and matches my purpose. But somehow I'd need to import my .yml translations into the translation database, after that I'd like to get rid of the original translation files. Currently the yml translations somehow aren't shown.

So, we agree with your basic scenario (didn't even thought about  the production caching of the last step that way).

TL;DR; Will give feedback on the rest of your comment later today.

**Comment by Pierstoval** [comment]
> You're right about this, and matches my purpose. But somehow I'd need to import my .yml translations into the translation database, after that I'd like to get rid of the original translation files. Currently the yml translations somehow aren't shown.

This means that the command must scan every bundle, merge all catalogues, and re-export all translations in the database.

> So, we agree with your basic scenario (didn't even thought about the production caching of the last step that way).
 
It's also done in debug mode, that's why adding or editing a new translation files implies the need of a clear cache.

**Comment by Rvanlaak** [comment]
With importing the yml files I meant the files we created ourselves, not any other bundle from the vendor directory ;) My point is that currently it does not read those translations, but that could be a configuration problem.

**Comment by Rvanlaak** [comment]
> It would need Timestampable and Blameable behaviors, or at least rewrite them, and I don't either want to rewrite or add a dependency to this bundle, it would make it more complex.

The bundle already relies on [`doctrine/*` and `symfony/symfony` and `sensio/framework-extra-bundle` ](https://github.com/Orbitale/TranslationBundle/blob/master/composer.json#L11-L14) so most people will already have those. Would be a quick win if you ask me :+1:

**Comment by Pierstoval** [comment]
Yep it depends on symfony and all the deps you tell, but it's mostly because it's intended to be used with the Symfony Standard Edition, in which these bundles are almost mandatory :)

**Comment by Rvanlaak** [comment]
By which I actually tried to say that changing the translation it's `dateModified` in a `LifecycleEvent` is two lines of code ;) ghehe

**Comment by Rvanlaak** [comment]
Regarding overriding the translator, maybe @adrianolek can help with that. It seems [his `TranslationListener` does the persistence magic](https://github.com/adrianolek/AOTranslationBundle/blob/master/EventListener/TranslationListener.php), and he [also overrides the `Translator`](https://github.com/adrianolek/AOTranslationBundle/blob/master/Translation/Translator.php).

**Comment by Pierstoval** [comment]
Thanks for all the tips, I currently have a hot bunch of work to do in the next weeks, in my job and in my different side-projects, and I'll be unavailable during two weeks in May. I'll watch over this all after that :)

**Comment by Rvanlaak** [comment]
Symfony 2.7 has been released, and the improvements to the Translation components look really nice! How is your side-project going? ;)

**Comment by Pierstoval** [comment]
Actually I just came back from my trip to England, I got PLENTY of work to do, but don't worry, I will work on this! I think I might need to simplify the whole package, but as I want it 2.3-compatible, it's gonna be a tough work. 
I think I'm gonna need to make a specific branch which will be 2.7+ only, because the translator has many cool features since 2.6 and 2.7

**Comment by Pierstoval** [comment]
Closing because bundle is abandoned anyway

**Comment by rvanlaak** [comment]
👍
